pkg/api/types.go

```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
 	Items []ReplicationController `json:"items"`
 }
 
+// DisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
+type DisruptionBudget struct {
+	// Specification of the desired behavior of the DisruptionBudget.
+	Spec DisruptionBudgetSpec `json:"spec,omitempty"`
+	// Most recently observed status of the DisruptionBudget.
+	Status DisruptionBudgetStatus `json:"status,omitempty"`
+}
+
+// DisruptionBudgetSpec is a description of a DisruptionBudget.
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits   []RateLimit `json:"rateLimits,omitempty"`
+
+	MinInstances MinInstances `json:"minInstances"`
+}
+
+type MinInstances struct {
+	// The minimum number of pods that must be available simultaneously.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MinAvailable intstr.IntOrString `json:"minAvailable,,omitempty"`
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	MinReadySeconds int `json:"minReadySeconds,omitempty"`
+}
+
+// DisruptionBudgetStatus represents information about the status of a DisruptionBudget. Status may trail the actual
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// The current disruption rate that the collection of pods is experiencing.
```

Comment:
Actually I guess this list should just be insertion timestamp. The thing you have now works fine if there is only one RateLimit, but if there are multiple RateLimits then you would need multiple arrays here. I don't remember anymore why I suggested to make these entries be insertion timestamp + RateLimit.Duration. I think just insertion timestamp is better. So I would change this to
```
// Array of timestamps of recent disruptions corresponding to this DisruptionBudget.
DisruptionInfo []unversioned.Time `json:disruptionInfo,omitempty"
```